src/ssr-template-helper.js

```diff
@@ -0,0 +1,122 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Capability} from './service/viewer-impl';
+import {dict, map} from './utils/object';
+import {iterateCursor} from './dom';
+
+/** The attributes we allow to be sent to the viewer. */
+const ATTRS_TO_SEND_TO_VIEWER = {
+  'amp-list': ['src', 'single-item', 'max-items'],
+  'amp-form': ['action-xhr'],
+};
+
+/**
+ * Helper, that manages the proxying of template rendering to the viewer.
+ */
+export class SsrTemplateHelper {
+
+  /**
+   * @param {string} sourceComponent
+   * @param {!./service/viewer-impl.Viewer} viewer
+   * @param {!./service/template-impl.Templates} templates
+   */
+  constructor(sourceComponent, viewer, templates) {
+
+    /** @private @const */
+    this.viewer_ = viewer;
+
+    /** @private @const */
+    this.templates_ = templates;
+
+    /** @private @const {!XMLSerializer} */
+    this.xmls_ = new XMLSerializer();
+
+    /** @private @const */
+    this.sourceComponent_ = sourceComponent;
+  }
+
+  /**
+   * Whether the viewer can render templates.
+   * @return {boolean}
+   */
+  isSupported() {
+    return this.viewer_.canRenderTemplates();
+  }
+
+  /**
+   * Proxies xhr and template rendering to the viewer and renders
+   * the response.
+   * @param {!Element} element
+   * return {!Promise<{data:{?JsonObject|string|undefined}}>}
+   */
+  fetchAndRenderTemplate(element) {
+    const inputsAsJson = this.getElementInputsAsJson_(element);
```

Comment:
This is wasted work for `amp-list` right?

Reply:
it will be if we decide to use the old logic and use that in order to support bindings in ssr templates. For now this is being used in amp-list and bindings are not supported but I will revisit in my next PR, whether we can update the bindings on the received template response from the viewer, as I haven't explored that as a per item binding update is what amp-bind seems to be doing.